src/Orchard/Environment/ShellBuilders/ShellContainerFactory.cs

```diff
@@ -56,50 +67,150 @@ public class ShellContainerFactory : IShellContainerFactory {
+                            else {
+                                // This item is not a decorator.
+                                // We need to add it to the list of concrete implementations. This allows us to know the names of the implementations that need to be decorated should a decorator for this interface exist.
+                                AddConcreteRegistrationName(registrationName, interfaceType, itemToBeRegistered.Item.Type);
                             }
                         }
 
-                        if (typeof(IEventHandler).IsAssignableFrom(item.Type)) {
-                            var interfaces = item.Type.GetInterfaces();
+                        if (itemToBeRegistered.IsEventHandler) {
+                            // var interfaces = itemToBeRegistered.Item.Type.GetInterfaces();
+                            var interfaces = itemToBeRegistered.InterfaceTypes;
                             foreach (var interfaceType in interfaces) {
 
                                 // register named instance for each interface, for efficient filtering inside event bus
                                 // IEventHandler derived classes only
-                                if (interfaceType.GetInterface(typeof (IEventHandler).Name) != null) {
+                                if (interfaceType.GetInterface(typeof(IEventHandler).Name) != null) {
                                     registration = registration.Named<IEventHandler>(interfaceType.Name);
                                 }
                             }
                         }
 
-                        foreach (var parameter in item.Parameters) {
+                        foreach (var parameter in itemToBeRegistered.Item.Parameters) {
                             registration = registration
                                 .WithParameter(parameter.Name, parameter.Value)
                                 .WithProperty(parameter.Name, parameter.Value);
                         }
                     }
 
+                    foreach (var decorator in decorators) {
+                        // We need to ensure that there is an implementation of this service that can be decorated
+                        if (!_concreteRegistrationNames.ContainsKey(decorator.InterfaceType) || _concreteRegistrationNames[decorator.InterfaceType] == null || !_concreteRegistrationNames[decorator.InterfaceType].Any()) {
+                            var exception = new OrchardFatalException(T("The only registered implementations of `{0}` are decorators. In order to avoid circular dependenices, there must be at least one implementation that is not marked with the `OrchardDecorator` attribute.", decorator.InterfaceType.FullName));
+                            Logger.Fatal(exception, "Could not complete dependency registration as a circular dependency chain has been found.");
+
+                            throw exception;
```

Comment:
Is throwing a fatal exception correct in this instance? Or should we just log and move on?